src/morphosamplers/sampler.py

```diff
@@ -0,0 +1,223 @@
+"""Tools for image resampling."""
+
+from typing import Tuple
+
+import numpy as np
+from scipy.ndimage import map_coordinates
+from scipy.spatial.transform import Rotation
+
+from .spline import Spline3D
+
+
+def generate_3D_grid(
+    grid_shape: Tuple[int, int, int] = (10, 10, 10),
+    grid_spacing: Tuple[float, float, float] = (1, 1, 1),
+    squeeze: bool = True,
+) -> np.ndarray:
+    """
+    Generate a 3D sampling grid with specified shape and spacing.
+
+    The grid generated is centered on the origin, has shape (m, n, p, 3) for
+    grid_shape (m, n, p), and spacing grid_spacing between neighboring points.
+
+    Parameters
+    ----------
+    grid_shape : Tuple[int, int, int]
+        The number of grid points along each axis.
+    grid_spacing : Tuple[float, float, float]
+        Spacing between points in the sampling grid.
+    squeeze: bool
+        Discard degenerate dimensions in the output.
+
+    Returns
+    -------
+    np.ndarray
+        Coordinate of points forming the 3D grid.
+    """
+    # create indices for x and y with correct spacing
+    coords = []
+    for dim in range(3):
+        ln = grid_shape[dim]
+        spacing = grid_spacing[dim]
+        if ln != 1:
+            coord = np.linspace(-0.5, 0.5, ln) * (ln - 1) * spacing
+        else:
+            coord = 0
+        coords.append(coord)
+    # convert to stack form (x, y, z, 3)
+    grid = np.stack(np.meshgrid(*coords, indexing="ij"), axis=3)
```

Comment:
This could be replaced with below which feels a little clearer - I like to keep the names I use for image dimensions separate from the names I use for coordinates which index into those dimensions

```suggestion
    grid = np.indices(grid_shape)
    grid = einops.rearrange(grid, 'zyx d h w -> d h w zyx')
    grid -= np.array(grid_shape) / 2
    grid *= spacing
```

```

Reply:
Awesome! But it's off by 0.5, will fix.